src/bindfs.c

```diff
@@ -998,15 +1011,78 @@ static int bindfs_access(const char *path, int wants)
         : -EACCES;
 }
 
+/**
+ * Perform an access check against the dirname of the given path.
+ *
+ * @param path The path to get the dirname of and performe the check against.
+ * @param wants The bitwise-inclusive OR of the access permissions to be
+ *     checked (R_OK, W_OK, X_OK).
+ *
+ * @return 0 if the dirname has the requested permissions, -EACCES otherwise.
+ */
+static int dirname_access_check(const char *path, int wants) {
```

Comment:
I assume you're referring to [`bindfs_link`](https://github.com/vimist/bindfs/blob/acl_support/src/bindfs.c#L1411)?
Just going off the `link(2)` man page where it says `EACCES` is returned when:

> Write access to the directory containing `newpath` is  denied, or search permission is denied for one of the directories in the path prefix of `oldpath` or `newpath`. (See also path_resolution(7).)

This makes sense to me, as you wouldn't need write access to the `oldpath`, as you're not changing anything there, so I _think_ they do unfortunately need to be separate to accommodate this.

Reply:
I actually meant the one in `bindfs_rename` that you already changed.

To clarify: you do definitely need both functions because sometimes you want to call `check_path_search_perms` without `dirname_access_check`, but if every call to `dirname_access_check` is accompanied by a call to `check_path_search_perms`, then `dirname_access_check` could instead contain the call to `check_path_search_perms`.

In the case you quoted, you'd then call only `dirname_access_check` on `newpath` and only `check_path_search_perms` on `oldpath`.